driver-core/src/main/com/mongodb/internal/Timeout.java

```diff
@@ -0,0 +1,256 @@
+     *         </ol>
+     *     </li>
+     *     <li>
+     *         0 means either of the following
+     *         <ol>
+     *             <li>0 nanoseconds has elapsed;</li>
+     *             <li>(n + 1) * 2<sup>63</sup><sup>(*)</sup> nanoseconds has elapsed,
+     *             in which case the timeout has expired.</li>
+     *         </ol>
+     *         Since it is impossible to differentiate the former from the latter, and the former is much more likely to happen in practice,
+     *         this class interprets 0 value as 0 elapsed nanoseconds.
+     *     </li>
+     *     <li>
+     *         A positive value means either of the following
+     *         <ol>
+     *             <li>this exact number of nanoseconds has elapsed;</li>
+     *             <li>((n + 1) * 2<sup>63</sup>; (n + 2) * 2<sup>63</sup> - 1]<sup>(*)</sup> nanoseconds has elapsed,
+     *             in which case the timeout has expired.</li>
+     *         </ol>
+     *         Since it is impossible to differentiate the former from the latter, and the former is much more likely to happen in practice,
+     *         this class interprets a positive value as the exact number of elapsed nanoseconds.
+     *     </li>
+     * </ul>
+     * <hr>
+     * <sup>(*)</sup> n is positive and odd.
+     */
+    private long elapsedNanos(final long currentNanos) {
+        assertFalse(isInfinite() || isImmediate());
+        return currentNanos - startNanos;
+    }
+
+    /**
+     * Is package-access for the purpose of testing and must not be used for any other purpose outside of this class.
+     * <p>
+     * Returns 0 or a positive value.
+     * 0 means that the timeout has expired.
+     * <p>
+     * Must not be called on {@linkplain #isInfinite() infinite} timeouts.
+     */
+    long remainingNanos(final long currentNanos) {
+        assertFalse(isInfinite());
+        if (isImmediate()) {
+            return 0;
+        } else {
+            long elapsedNanos = elapsedNanos(currentNanos);
+            return elapsedNanos < 0 ? 0 : Math.max(0, durationNanos - elapsedNanos);
+        }
+    }
+
+    /**
+     * Returns 0 or a positive value.
+     * Use {@link #expired(long)} to check if the returned value signifies that a timeout is expired.
+     *
+     * @throws UnsupportedOperationException If the timeout is {@linkplain #isInfinite() infinite}.
+     * @see #remainingNanosOrInfinite()
+     */
+    public long remainingNanos() throws UnsupportedOperationException {
+        if (isInfinite()) {
+            throw new UnsupportedOperationException();
+        }
+        return remainingNanos(System.nanoTime());
+    }
+
+    /**
+     * Returns a negative value for {@linkplain #isInfinite() infinite} timeouts, otherwise 0 or a positive value.
+     * Use {@link #expired(long)} to check if the returned value signifies that a timeout is expired.
+     *
+     * @see #remainingNanos()
+     */
+    public long remainingNanosOrInfinite() {
```

Comment:
Suggestion: As we use Milliseconds internally for all our timeouts and we use : `setSomeTimeout(long timeout, TimeUnit timeUnit)` as the signature perhaps this method could take a `TimeUnit`.

What do you think?

Reply:
I initially considered having the `remaining(TimeUnit)` method, but having it is not as straightforward as it appears:

1. Converting from nanoseconds to a more coarse units cannot generally be done without losing precision. `TimeUnit.convert` does this by trimming the fractional part: `MILLISECONDS.convert(999_999, TimeUnit.NANOSECONDS)` is 0.
2. As a result of the item 1, the method `remaining(TimeUnit)` may return values smaller than the actual remaining time, and result in either reducing the timeout when passed to methods that accept milliseconds (e.g., `Object.wait​(long timeoutMillis)`), or having a similar effect when passed to `Timeout.expired(long remaining)` (if we have `remaining(TimeUnit)`, the `expired(long)` method has to accept any time unit).
3. Since we are not in the business of writing real-time code (surprisingly, that's a [thing in Java](https://www.jcp.org/en/jsr/detail?id=1), and a [new version of RTSJ is even being developed](https://www.aicas.com/wp/products-services/technology/#rtsj)), we can (and practically do) miss the timeouts (return later)<sup>1</sup>, however, we must not return before a timeout will have expired. The item 2 gets in a way of this requirement.

So I decided to not have the `remaining(TimeUnit)` method. However, now that you asked for it, I thought again, and realized that by not providing the method, I am simply shifting the burden to someone who will have to pass remaining time from `Timeout` to a method that accepts milliseconds. And that will likely be done by converting and trimming the fractional part, which will lead to us potentially timing out early, thus violating the timeout contract.

The only way I see to have coarsening conversion and not timeout early is to round up the result. This is what `RoundingMode.UP`/`RoundingMode.CEILING` do (since we do not need to round negative numbers, these two are equivalent). However, since I do not want to re-write `TimeUnit.convert` from scratch, I cannot just use the rounding mode and need to use a different approach.

---
<sup>1</sup>If one rigorously thinks about what a timeout is, then only the following seems to be achievable without additional guarantees (which do not exist in Java SE, and I am not even sure what they are): a method timeout means that a method will try to accomplish the task for **at least** the specified amount of time, and if neither succeeds nor detects an exceptional situation, the method returns at **some instant after** that time elapses.

Reply:
Done.